### PR TITLE
Elect leader api(KIP-460) implemented

### PR DESCRIPTION
**Pull request by PratRanj07**
Implemented the ElectLeaders API and wrote unit tests for it

### CODE REVIEW

File: kafka/adminapi.go [hunk 1]
```diff
@@ -1048,6 +1055,50 @@ type ListOffsetsResult struct {
...
+	return ElectLeadersRequest{
+		electionType: electionType,
+		partitions:   partitions,
+	}
+}
+
+// ElectLeadersResult holds the result of the election performed
+type ElectLeadersResult struct {
+	// TopicPartitions for which election has been performed and specific error if any
+	// that occured while election in the specific TopicPartition.
```
Comment (milindl) [inline kafka/adminapi.go:1098]:
```suggestion
	// TopicPartitions for which election has been performed and the per-partition error, if any
	// that occurred while running the election for the specific TopicPartition.
```

File: kafka/adminapi.go [hunk 2]
```diff
@@ -3516,6 +3594,78 @@ func (a *AdminClient) DeleteRecords(ctx context.Context,
...
+// Parameters:
+//   - `ctx` - context with the maximum amount of time to block, or nil for
+//     indefinite.
+//   - `electLeaderRequest` - ElectLeadersRequest containing the election type
+//     and the partitions to elect leaders for.
+//   - `options` - ElectLeadersAdminOption options.
+//
+// Returns ElectLeadersResult, which contains a slice of TopicPartitions containing the partitions for which the leader election was performed.
+// User has to check all the elements of the result to check any error occured per partition.
+// err returns any top level level that occured during the operation.
```
Comment (milindl) [inline kafka/adminapi.go:3608]:
```suggestion
// Individual TopicPartitions inside the ElectLeadersResult should be checked for errors.
// Additionally, an error that is not nil for client-level errors is returned.
```

File: kafka/adminapi_test.go [hunk 3]
```diff
@@ -918,6 +918,80 @@ func testAdminAPIsDeleteRecords(what string, a *AdminClient, expDuration time.Du
...
+	_, err = a.ElectLeaders(ctx, electLeadersRequestUnclean, SetAdminRequestTimeout(-1))
+	if err == nil || err.(Error).Code() != ErrInvalidArg {
+		t.Fatalf("Expected ErrInvalidArg, not %v", err)
+	}
+
+	for _, options := range [][]ElectLeadersAdminOption{
+		{},
+		{SetAdminRequestTimeout(time.Second)},
+	} {
+		// empty list should fail
```
Comment (milindl) [inline kafka/adminapi_test.go:964]: Change this after the librdkafka change, so the empty list fails, but a `nil` list times out instead

File: examples/admin_elect_leaders/admin_elect_leaders.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+		i += 2
+		index++
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+
+	res, err := ac.ElectLeaders(ctx, kafka.NewElectLeadersRequest(electionType, topicPartitionList))
+	if err != nil {
+		kafkaErr, ok := err.(kafka.Error) // Type assertion
```
Comment (milindl) [inline examples/admin_elect_leaders/admin_elect_leaders.go:88]:
```suggestion
		kafkaErr, ok := err.(kafka.Error)
```

**Review by milindl — CHANGES_REQUESTED** [review]
Looks pretty good

Please add an integration test

File: kafka/adminapi_test.go (hunk 3 above)
Comment (PratRanj07) [reply kafka/adminapi_test.go]: Actually after the librdkafka change, both empty list and nil partitions is valid.
If there is an empty list in input, we are getting an empty list in the output and for nil, all partitions.

File: examples/README.md [hunk 5]
```diff
@@ -28,6 +28,8 @@ Examples
 
   [admin_describe_topics](admin_describe_topics) - Describe topics
 
+  [admin_elect_leaders](admin_elect_leaders) - Perform Preferred/Unclean elections for the specified Topic Partitions
```
Comment (emasab) [inline examples/README.md:31]:
```suggestion
  [admin_elect_leaders](admin_elect_leaders) - Perform Preferred/Unclean elections for the specified partitions or for all of them
```

File: examples/README.md [hunk 6]
```diff
@@ -86,6 +88,7 @@ Examples
 
   [transactions_example](transactions_example) - Showcasing a transactional consume-process-produce application
 
+
```
Comment (emasab) [inline examples/README.md:91]: Remove this newline addition

File: examples/admin_elect_leaders/admin_elect_leaders.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+	res, err := ac.ElectLeaders(ctx, kafka.NewElectLeadersRequest(electionType, topicPartitionList))
+	if err != nil {
+		kafkaErr, ok := err.(kafka.Error)
+		if ok && kafkaErr.Code() != kafka.ErrNoError {
+			fmt.Printf("Failed to elect Leaders: %s\n", err)
+			os.Exit(1)
+		}
+	}
+
+	fmt.Printf("ElectLeaders result: %+v\n", res)
```
Comment (emasab) [inline examples/admin_elect_leaders/admin_elect_leaders.go:96]: In the examples it's better to print the result manually to see how it needs to be handled, and to verify it's accessible.
```suggestion
	fmt.Printf("ElectLeaders result contains %d partition(s):\n",
		len(res.TopicPartitions))
	for _, topicPartition := range res.TopicPartitions {
		kafkaErr, ok := topicPartition.Error.(kafka.Error)
		if !ok {
			fmt.Printf("Topic: %s, Partition: %d - Unexpected error type: %s\n",
				*topicPartition.Topic, topicPartition.Partition, err)
			continue
		}

		if kafkaErr.Code() == kafka.ErrNoError {
			fmt.Printf("Topic: %s, Partition: %d - Success\n",
				*topicPartition.Topic, topicPartition.Partition)
		} else {
			fmt.Printf("Topic: %s, Partition: %d - Failed: %s\n",
				*topicPartition.Topic, topicPartition.Partition, kafkaErr.String())
		}
	}
```

File: kafka/adminapi.go [hunk 8]
```diff
@@ -145,6 +145,13 @@ error_by_idx(const rd_kafka_error_t **errors, size_t cnt, size_t idx) {
 		return NULL;
 	return errors[idx];
 }
+
+static const rd_kafka_topic_partition_result_t *
+TopicPartitionResult_by_idx(const rd_kafka_topic_partition_result_t **result, size_t cnt, size_t idx) {
+	if (idx >= cnt)
+		return NULL;
+	return result[idx];
+}
```
Comment (emasab) [inline kafka/adminapi.go:154]:
```suggestion

static const rd_kafka_topic_partition_result_t *
TopicPartitionResult_by_idx(const rd_kafka_topic_partition_result_t **results, size_t cnt, size_t idx) {
	if (idx >= cnt)
		return NULL;
	return results[idx];
}
```

File: kafka/adminapi.go [hunk 9]
```diff
@@ -1048,6 +1055,54 @@ type ListOffsetsResult struct {
...
+		electionType: electionType,
+		partitions:   partitions,
+	}
+}
+
+// ElectLeadersResult holds the result of the election performed
+type ElectLeadersResult struct {
+	// TopicPartitions for which election has been performed and the per-partition error, if any
+	// that occurred while running the election for the specific TopicPartition.
+	topicPartitions []TopicPartition
```
Comment (emasab) [inline kafka/adminapi.go:1103]: This must be made public
```suggestion
	TopicPartitions []TopicPartition
```

File: kafka/adminapi.go [hunk 10]
```diff
@@ -3516,6 +3592,74 @@ func (a *AdminClient) DeleteRecords(ctx context.Context,
...
+	err = a.verifyClient()
+	if err != nil {
+		return result, err
+	}
+
+	cTopicPartitions := newCPartsFromTopicPartitions(electLeaderRequest.partitions)
+	defer C.rd_kafka_topic_partition_list_destroy(cTopicPartitions)
+
+	cElectLeadersRequest := C.rd_kafka_ElectLeaders_new(C.rd_kafka_ElectionType_t(electLeaderRequest.electionType), cTopicPartitions)
+	defer C.rd_kafka_ElectLeaders_destroy(cElectLeadersRequest)
```
Comment (emasab) [inline kafka/adminapi.go:3620]: This way we handle the nil case differently from the empty case
```suggestion
        var cTopicPartitions *C.rd_kafka_topic_partition_list_t
	if electLeaderRequest.partitions != nil {
		cTopicPartitions = newCPartsFromTopicPartitions(electLeaderRequest.partitions)
		defer C.rd_kafka_topic_partition_list_destroy(cTopicPartitions)
	}
```

File: kafka/adminapi.go [hunk 11]
```diff
@@ -3516,6 +3592,74 @@ func (a *AdminClient) DeleteRecords(ctx context.Context,
...
+		return result, err
+	}
+	defer C.rd_kafka_event_destroy(rkev)
+
+	cResEvent := C.rd_kafka_event_ElectLeaders_result(rkev)
+	cRes := C.rd_kafka_ElectLeaders_result(cResEvent)
+
+	var cResponseSize C.size_t
+	cResultPartitions := C.rd_kafka_ElectLeadersResult_partitions(cRes, &cResponseSize)
+	result.topicPartitions = newTopicPartitionsFromCTopicPartitionResult(cResultPartitions, cResponseSize)
```
Comment (emasab) [inline kafka/adminapi.go:3658]: After librdkafka PR for removing the duplicate result it becomes:
```suggestion
	cRes := C.rd_kafka_event_ElectLeaders_result(rkev)

	var cResponseSize C.size_t
	cResultPartitions := C.rd_kafka_ElectLeaders_result_partitions(cRes, &cResponseSize)
	result.TopicPartitions = newTopicPartitionsFromCTopicPartitionResult(cResultPartitions, cResponseSize)

```

**Review by emasab — CHANGES_REQUESTED** [review]
Great PR Pratyush! A few comments:

**Pull request merged**